Assets/MixedRealityToolkit.Tests/PlayModeTests/BaseCursorTests.cs

```diff
@@ -0,0 +1,339 @@
+
+            var collider = cube.GetComponentInChildren<Collider>();
+            Assert.IsNotNull(collider);
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            Object.Destroy(cube);
+            TestUtilities.ShutdownMixedRealityToolkit();
+        }
+
+        private void VerifyCursorStateFromPointers(IEnumerable<IMixedRealityPointer> pointers, CursorStateEnum state)
+        {
+            Assert.NotZero(pointers.ToReadOnlyCollection().Count);
+            foreach(var pointer in pointers)
+            {
+                VerifyCursorState(pointer.BaseCursor, state);
+            }
+        }
+
+        private void VerifyCursorState(IMixedRealityCursor cursor, CursorStateEnum state)
+        {
+            var baseCursor = cursor as BaseCursor;
+            Assert.IsNotNull(baseCursor);
+            Assert.AreEqual(state, baseCursor.CursorState);
+        }
+
+        private void VerifyCursorContextFromPointers(IEnumerable<IMixedRealityPointer> pointers, CursorContextEnum context)
+        {
+            Assert.NotZero(pointers.ToReadOnlyCollection().Count);
+            foreach (var pointer in pointers)
+            {
+                VerifyCursorContext(pointer.BaseCursor, context);
+            }
+        }
+
+        private void VerifyCursorContext(IMixedRealityCursor cursor, CursorContextEnum context)
+        {
+            var baseCursor = cursor as BaseCursor;
+            Assert.IsNotNull(baseCursor);
+            Assert.AreEqual(context, baseCursor.CursorContext);
+        }
+
+        [UnityTest]
+        public IEnumerator ArticulatedCursorState()
+        {
+            var inputSystem = PlayModeTestUtilities.GetInputSystem();
+
+            // Show hand far enough from the test collider so the cursor is not on it
+            var hand = new TestHand(Handedness.Right);
+            Vector3 offObjectPos = new Vector3(0.05f, 0, 1.0f);
+            yield return hand.Show(offObjectPos);
+            VerifyCursorStateFromPointers(inputSystem.FocusProvider.GetPointers<ShellHandRayPointer>(), CursorStateEnum.Interact);
+
+            // Move hand closer to the collider so the cursor is on it
+            Vector3 onObjectPos = new Vector3(0.05f, 0, 1.5f);
+            yield return hand.MoveTo(onObjectPos, numFramesPerMove);
+            VerifyCursorStateFromPointers(inputSystem.FocusProvider.GetPointers<ShellHandRayPointer>(), CursorStateEnum.InteractHover);
+
+            // Trigger pinch
+            yield return hand.SetGesture(ArticulatedHandPose.GestureId.Pinch);
+            VerifyCursorStateFromPointers(inputSystem.FocusProvider.GetPointers<ShellHandRayPointer>(), CursorStateEnum.Select);
+
+            // Release pinch
+            yield return hand.SetGesture(ArticulatedHandPose.GestureId.Open);
+            VerifyCursorStateFromPointers(inputSystem.FocusProvider.GetPointers<ShellHandRayPointer>(), CursorStateEnum.Release);
```

Comment:
Do I understand correctly, that Release and InteractHover _must_ come on different frames? If that's the case, yield WaitForFixedUpdate should probably be just `yield return null`.

Reply:
I'm not so sure on the exact behaviour, but when I change this to `yield return null` it fails 50% of the time.

Reply:
Interesting, ok, thanks for clarification.